src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,189 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    public class Node<T> {
+        protected T value;
+        protected Node<T> prev;
+        protected Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private int size = 0;
+    private Node<T> first;
+    private Node<T> last;
+
+    private void checkPositionIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    private void checkElementIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    Node<T> node(int index) {
+        // assert isElementIndex(index);
+
+        if (index < (size >> 1)) {
+            Node<T> x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        } else {
+            Node<T> x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x;
+        }
+    }
+
+    private void linkLast(T value) {
+        final Node<T> l = last;
+        final Node<T> newNode = new Node<>(l, value, null);
+        last = newNode;
+        if (l == null) {
+            first = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkBefore(T value, Node<T> node) {
+        final Node<T> previous = node.prev;
+        final Node<T> newNode = new Node<>(previous, value, node);
+        node.prev = newNode;
+        if (previous == null) {
+            first = newNode;
+        } else {
+            previous.next = newNode;
+        }
+        size++;
+    }
+
+    T unlink(Node<T> removedElement) {
+        final T element = removedElement.value;
+        final Node<T> next = removedElement.next;
+        final Node<T> prev = removedElement.prev;
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            removedElement.prev = null;
```

Comment:
line 84, 90 and 92 are not needed, since we remove the ability to get into this node at line 83 and 89